pallets/dapp-staking-v3/src/types.rs

```diff
@@ -638,4 +813,452 @@ impl EraInfo {
+    pub fn unstake(
+        &mut self,
+        amount: Balance,
+        period_info: PeriodInfo,
+        era: EraNumber,
+    ) -> Result<(), ()> {
+        // Defensive check to ensure we don't end up in a corrupted state. Should never happen.
+        if let Some(last_element) = self.0.last() {
+            // It's possible last element refers to the upcoming era, hence the "-1" on the 'era'.
+            if last_element.era().saturating_sub(1) > era
+                || last_element.period() > period_info.number
+            {
+                return Err(());
+            }
+        } else {
+            // Vector is empty, should never happen.
+            return Err(());
+        }
+
+        // 1st step - remove the last element IFF it's for the next era.
+        // Unstake the requested amount from it.
+        let last_era_info = match self.0.last() {
+            Some(last_element) if last_element.era() == era.saturating_add(1) => {
+                let mut last_element = *last_element;
+                last_element.unstake(amount, period_info.period_type);
+                let _ = self.0.pop();
+                Some(last_element)
+            }
+            _ => None,
+        };
+
+        // 2nd step - 3 options:
+        // 1. - last element has a matching era so we just update it.
+        // 2. - last element has a past era and matching period, so we'll create a new entry based on it.
+        // 3. - last element has a past era and past period, meaning it's invalid.
+        let second_last_era_info = if let Some(last_element) = self.0.last_mut() {
+            if last_element.era() == era {
+                last_element.unstake(amount, period_info.period_type);
+                None
+            } else if last_element.period() == period_info.number {
+                let mut new_entry = *last_element;
+                new_entry.unstake(amount, period_info.period_type);
+                new_entry.era = era;
+                Some(new_entry)
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+
+        // 3rd step - push the new entries, if they exist.
+        if let Some(info) = second_last_era_info {
+            self.prune();
+            self.0.try_push(info).map_err(|_| ())?;
+        }
+        if let Some(info) = last_era_info {
+            self.prune();
+            self.0.try_push(info).map_err(|_| ())?;
+        }
+
+        Ok(())
+    }
+
+    /// Used to remove past entries, in case vector is full.
+    fn prune(&mut self) {
+        // Prune the oldest entry if we have more than the limit
+        if self.0.len() == STAKING_SERIES_HISTORY as usize {
+            // TODO: this can be perhaps optimized so we prune entries which are very old.
+            // However, this makes the code more complex & more error prone.
+            // If kept like this, we always make sure we cover the history, and we never exceed it.
```

Comment:
The only possible optimization would be reduced PoV (but entry is quite small anyhow).

The "drawback" of the current solution is that very old entries might need some time to get cleaned up. In reality, for active contracts this won't happen. And even if it does, the effect is minuscule and extremely situational.

E.g. image a contract being staked on in era `5`, then again staked on in era `6`, and once again in era `20`.
Realistically, we only need information for the last 3 eras, but using the current logic we'll keep all values inside the `Series`. Optimized solution would be to recognize that latest is `20`, and that we only need values up to `18`. Vector size would be reduced to 2, instead of 3.